src/node_api.cc

```diff
@@ -17,17 +16,23 @@
 #include <vector>
 #include "node_api.h"
 
-namespace v8impl {
-
-//=== Conversion between V8 Isolate and napi_env ==========================
+napi_status napi_set_last_error(napi_env env, napi_status error_code,
+                                uint32_t engine_error_code = 0,
+                                void* engine_reserved = nullptr);
+void napi_clear_last_error(napi_env env);
 
-napi_env JsEnvFromV8Isolate(v8::Isolate* isolate) {
-  return reinterpret_cast<napi_env>(isolate);
-}
+class napi_env__ {
+ public:
+  explicit napi_env__(v8::Isolate* _isolate): isolate(_isolate), last_error() {}
+  ~napi_env__() {
+    last_exception.Reset();
+  }
+  v8::Isolate* isolate;
+  v8::Persistent<v8::Value> last_exception;
+  napi_extended_error_info last_error;
+};
```

Comment:
I'm not sure what we could provide to implementations. For example, `last_exception` only exists in the V8 implementation, because in ChakraCore `napi_get_and_clear_last_exception()` goes straight through to JSRT, so we likely won't need to store any last exception on the `env` in the ChakraCore implementation.

Of course, we could come up with an API for implementors, wherein we might expose the minimal definition of `napi_env__` for them to extend, but I think that's a substantial exercise.